lib/Doctrine/ORM/Mapping/AssociationMapping.php

```diff
@@ -0,0 +1,226 @@
+     *                                   is. If it is a simple, unqualified
+     *                                   class name the namespace is assumed to
+     *                                   be the same as the namespace of the
+     *                                   source entity.
+     */
+    final public function __construct(
+        public readonly int $type,
+        public string $fieldName,
+        public string $sourceEntity,
+        public string $targetEntity,
+    ) {
+    }
+
+    /** @param array{fieldName: string, sourceEntity: class-string, targetEntity: class-string, type: int, joinColumns?: mixed[]|null, joinTable?: mixed[]|null, ...} $mappingArray */
+    public static function fromMappingArray(array $mappingArray): self
+    {
+        unset($mappingArray['isOwningSide']);
+        $mapping = new static(
+            $mappingArray['type'],
+            $mappingArray['fieldName'],
+            $mappingArray['sourceEntity'],
+            $mappingArray['targetEntity'],
+        );
+        foreach ($mappingArray as $key => $value) {
+            if (in_array($key, ['type', 'fieldName', 'sourceEntity', 'targetEntity'])) {
+                continue;
+            }
+
+            if ($key === 'joinTable') {
+                assert($mapping instanceof ManyToManyAssociationMapping);
+
+                if ($value === [] || $value === null) {
+                    continue;
+                }
+
+                assert($mapping instanceof ManyToManyOwningSideMapping);
+
+                $mapping->joinTable = JoinTableMapping::fromMappingArray($value);
+
+                continue;
+            }
+
+            if (property_exists($mapping, $key)) {
+                $mapping->$key = $value;
+            } else {
+                throw new Exception('Unknown property ' . $key . ' on class ' . static::class);
+            }
+        }
+
+        return $mapping;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function offsetExists($offset): bool
+    {
+        return isset($this->$offset);
+    }
+
+    public function offsetGet($offset): mixed
+    {
+        if ($offset === 'isOwningSide') {
+            return $this->isOwningSide();
+        }
+
+        return $this->$offset;
+    }
+
+    /** @psalm-assert AssociationOwningSideMapping $this */
+    public function isOwningSide(): bool
+    {
+        return $this instanceof AssociationOwningSideMapping;
+    }
+
+    /**
+     * @psalm-assert-if-true ToOneAssociationMapping $this
+     * @psalm-assert-if-true AssociationOwningSideMapping $this
+     */
+    public function isToOneOwningSide(): bool
+    {
+        return $this instanceof ToOneAssociationMapping && $this->isOwningSide();
+    }
+
+    /** @psalm-assert-if-true ManyToManyOwningSideMapping $this */
+    public function isManyToManyOwningSide(): bool
+    {
+        return $this instanceof ManyToManyOwningSideMapping;
+    }
```

Comment:
@mpdude doing `instanceof` checks got old real fast, so I introduced the 3 methods above, and created https://github.com/phpstan/phpstan/issues/8904